### ISSUE TITLE
Implement contraty to url_to_fs

### ISSUE THREAD

**Issue by juliospain** [intake/filesystem_spec#538]
With url_to_fs function we can create a (potentially chained) url to a fs and relative path in this fs, which can later be used to open the file/folder or whatever.
Once we have this fs and url, it is impossible currently to obtain the original complete chained url as far as I know (or at least in a general way). Potential uses are logging, storing user settings like last recently used files, etc.

I think it would be good to have a function inside the file systems that after passing a url gives the "full url" (let's call it like that. For that all those file system that have target_protocol to store internally so a chained call can be made to those.
I have not talk about the target_options (user, password, etc...) but potentially should be included as well, and the correct display be handled in a general or case-by-case (I do not have an overview at the moment beyond the ftp protocol).

What do you think? If it is something that could be interesting in the baseline fsspec, I could work on it (as last time, at my own pace :P)

**Comment by martindurant** [comment]
Yes, I think this could be useful, and it need not necessarily be method, you could have a function `canonical_url(chained_filesystem, relative_url)`. Yes, you are quite right to point out `target_protocol`, but all parameters passed at instantiation time are stored in an instance: `storage_args` (tuple) and `storage_options` (dict for kwargs). That should be all the information you need.

**Comment by juliospain** [comment]
Fair enough, it would make sense since url_to_fs is not a method but a function.

Still some specialized code per implementation would be needed after reading a little bit more. The unchain function heavily uses the _get_kwargs_from_urls. I will think about writing a "_encode_url_from_kwargs" or something like that.

**Comment by martindurant** [comment]
Maybe it's not necessary to encode any kwargs back into the URL, but rather to return a full kwargs diction including them all.

**Comment by martindurant** [comment]
@juliospain , were you thinking to work on this?

**Comment by juliospain** [comment]
I have indeed start working on this, just to realize:
1. That indeed it is better not to encode username or passwords in the resulting URL (this is just noise, and potentially in the case of passwords not good)
2. There are many small things to take into account (protocol lists, cache, etc...)

This is where I stand so far...

```
    def canonical_url(fs, urlpath) -> str:
        protocol = fs.protocol
        protocol = protocol if isinstance(protocol, str) else protocol[0] # protocol list?
        target_options = fs.storage_options

        url = ""
        while protocol:
            
            _, urlpath = fsspec.core.split_protocol(urlpath)
            
            if protocol in {"blockcache", "filecache", "simplecache"}:
                url += "::" + f"{protocol}"
            elif protocol == default_protocol:
                url += "::" + f"{urlpath}"
            else:
                url += "::" + f"{protocol}://{urlpath}"
            
            urlpath = target_options.get('fo',None)
            protocol = target_options.get('target_protocol', None)
            if protocol:
                protocol = protocol if isinstance(protocol, str) else protocol[0] # protocol list?
            target_options = target_options.get('target_options', {})
            
        return url[2:] #strip initial ::
```
